src/components/input/SelectNext.tsx

```diff
@@ -103,38 +105,108 @@ function SelectOption<T>({
 
 SelectOption.displayName = 'SelectNext.Option';
 
-function useShouldDropUp(
+const LISTBOX_TOGGLE_GAP = '.25rem';
+
+type ListboxCSSProps =
+  | 'top'
+  | 'left'
+  | 'minWidth'
+  | 'marginBottom'
+  | 'bottom'
+  | 'marginTop';
+
+function useListboxPositioning(
   buttonRef: RefObject<HTMLElement | undefined>,
   listboxRef: RefObject<HTMLElement | null>,
   listboxOpen: boolean,
-): boolean {
-  const [shouldListboxDropUp, setShouldListboxDropUp] = useState(false);
+  asPopover: boolean,
+  right: boolean,
+) {
+  const adjustListboxPositioning = useCallback(() => {
+    const listboxEl = listboxRef.current;
+    const buttonEl = buttonRef.current;
 
-  useLayoutEffect(() => {
-    // Reset shouldListboxDropUp so that it does not affect calculations next
-    // time listbox opens
-    if (!buttonRef.current || !listboxRef.current || !listboxOpen) {
-      setShouldListboxDropUp(false);
-      return;
+    if (!buttonEl || !listboxEl || !listboxOpen) {
+      return () => {};
     }
 
+    /**
+     * Sets some CSS props in the listbox, and returns a cleanup function that
+     * resets them to their default values.
+     */
+    const setListboxCSSProps = (
+      props: Partial<Record<ListboxCSSProps, string>>,
+    ) => {
+      const entries = Object.entries(props) as Array<[ListboxCSSProps, string]>;
+      entries.forEach(([prop, value]) => (listboxEl.style[prop] = value));
+
+      return () => entries.map(([prop]) => (listboxEl.style[prop] = ''));
```

Comment:
We do, because the effect depends on `isListboxOpen`, so it runs every time the listbox is shown/hidden.

Without the cleanup, if the listbox was first slidden-up, then you scroll further down and click the toggle again, it keeps the previous styles which conflict with the new ones.

Basically, we reset the styles every time the listbox is closed, and set new ones when opened.